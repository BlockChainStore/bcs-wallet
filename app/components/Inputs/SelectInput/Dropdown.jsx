// @flow
import React from 'react'
import classNames from 'classnames'
import withClickOutside from 'react-click-outside'
import { noop } from 'lodash'

import styles from './Dropdown.scss'

declare type NodeOrNodes = React$Node | Array<React$Node>

type Props = {
  className: string,
  children: NodeOrNodes,
  open: boolean,
  onClose: Function,
  renderDropdown: Function
}

class Dropdown extends React.Component<Props> {
  static defaultProps = {
    open: false,
    onClose: noop,
    renderDropdown: noop
  }

  render = () => {
    return (
      <div className={classNames(styles.dropdown, this.props.className)}>
        <div className={styles.anchor}>
          {this.props.children}
        </div>
        {this.renderContent()}
      </div>
    )
  }

  renderContent = () => {
    if (this.props.open) {
      const DropdownContent = this.props.renderDropdown
      return <DropdownContent className={styles.content} />
    }
  }

  handleClickOutside = () => {
    this.props.onClose()
  }
}

export default withClickOutside(Dropdown)
