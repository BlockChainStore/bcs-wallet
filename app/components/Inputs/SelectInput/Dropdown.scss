.dropdown {
  position: relative;

  .content {
    position: absolute;
  }
}
