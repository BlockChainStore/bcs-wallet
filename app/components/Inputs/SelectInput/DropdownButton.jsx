// @flow
import React from 'react'
import classNames from 'classnames'
import ArrowDropDown from 'react-icons/lib/md/arrow-drop-down'

import styles from './DropdownButton.scss'

type Props = {
  className?: string,
  onToggle: Function
}

const DropdownButton = ({ className, onToggle }: Props) => (
  <div className={classNames(styles.dropdownButton, className)} onClick={() => onToggle()}>
    <ArrowDropDown />
  </div>
)

export default DropdownButton
