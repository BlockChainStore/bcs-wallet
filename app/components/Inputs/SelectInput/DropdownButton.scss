@import '../../../styles/variables';

.dropdownButton {
  display: flex;
  align-items: center;
  height: 100%;
  padding: 0 6px;
  background: $primary-color;
  color: $button-text;
  cursor: pointer;
}
