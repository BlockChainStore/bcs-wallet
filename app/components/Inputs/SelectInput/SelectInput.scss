@import '../../../styles/variables';

.selectInput {
  position: relative;

  .inputContainer {
    display: flex;
    background: $dark-input-color;

    .input {
      flex: 1 1 auto;
      box-sizing: border-box;
      width: 100%;
    }

    .afterInput {
      flex: 0 1 auto;
    }
  }

  .dropdown {
    box-sizing: border-box;
    width: 100%;
    max-height: 180px;
    overflow: auto;
    background: white;
    box-shadow: 0px 4px 4px 0px rgba(black, 0.2);
    z-index: 1;

    .dropdownItem {
      padding: 10px 20px;
      color: #444;
      font-size: 14px;
      cursor: pointer;

      &:hover {
        background: #eee;
      }
    }

    &.noSearchResults {
      padding: 10px 20px;
      color: #9e9e9e;
      text-align: center;
      font-size: 13px;
      font-style: italic;
    }
  }
}
