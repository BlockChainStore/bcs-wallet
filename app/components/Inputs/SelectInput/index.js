export { default } from './SelectInput'
export { default as DropdownButton } from './DropdownButton'
