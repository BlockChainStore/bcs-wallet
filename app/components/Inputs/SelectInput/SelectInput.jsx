// @flow
import React from 'react'
import classNames from 'classnames'
import { noop, omit, trim, includes, lowerCase } from 'lodash'

import Dropdown from './Dropdown'
import DropdownButton from './DropdownButton'
import styles from './SelectInput.scss'

type Props = {
  className?: string,
  value?: string,
  items: Array<any>,
  renderItem: Function,
  renderAfter: Function,
  getItemValue: Function,
  getSearchResults: Function,
  onFocus?: Function,
  onChange?: Function
}

type State = {
  open: boolean,
  search: string
}

const defaultRenderAfter = (props) => <DropdownButton {...props} />

const defaultItemValue = (item) => item

const defaultSearchResults = (items, term) => {
  return items.filter((item) => includes(lowerCase(item), lowerCase(term)))
}

export default class SelectInput extends React.Component<Props, State> {
  static defaultProps = {
    items: [],
    renderAfter: defaultRenderAfter,
    getItemValue: defaultItemValue,
    getSearchResults: defaultSearchResults,
    onFocus: noop,
    onChange: noop
  }

  state = {
    open: false,
    search: ''
  }

  componentWillReceiveProps = (nextProps: Props) => {
    if (nextProps.value !== this.props.value) {
      this.setState({ search: trim(nextProps.value) })
    }
  }

  render = () => {
    const passDownProps = omit(this.props, 'items', 'renderItem', 'renderAfter', 'getItemValue', 'getSearchResults')

    return (
      <Dropdown
        className={classNames(styles.selectInput, this.props.className)}
        open={this.state.open}
        onClose={this.handleClose}
        renderDropdown={this.renderDropdown}>
        <div className={styles.inputContainer}>
          <input
            {...passDownProps}
            className={styles.input}
            type='text'
            onFocus={this.handleFocus}
            onChange={this.handleChange}
          />
          <div className={styles.afterInput}>
            {this.props.renderAfter && this.props.renderAfter({ onToggle: this.handleToggle })}
          </div>
        </div>
      </Dropdown>
    )
  }

  renderDropdown = ({ className }: { className: string }) => {
    const items = this.getItems()
    const hasItems = items.length > 0
    const isSearch = this.state.search.length > 0 && this.props.items && this.props.items.length > 0

    if (hasItems) {
      return (
        <div className={classNames(styles.dropdown, className)}>
          {this.renderItems(items)}
        </div>
      )
    } else if (isSearch) {
      return (
        <div className={classNames(styles.dropdown, styles.noSearchResults, className)}>
          No search results.
        </div>
      )
    } else {
      return null
    }
  }

  renderItems = (items: Array<any>) => {
    return items.map((item) => {
      const renderItem = this.props.renderItem || this.renderItem
      return renderItem(item, { onSelect: this.generateSelectHandler(item) })
    })
  }

  renderItem = (item: Object, { onSelect }: { onSelect: Function }) => {
    const { getItemValue } = this.props
    return (
      <div className={styles.dropdownItem} key={getItemValue(item)} tabIndex={0} onClick={onSelect}>
        {item}
      </div>
    )
  }

  handleFocus = (event: Object, ...args: Array<any>) => {
    const { onFocus } = this.props
    event.persist()
    if (onFocus) {
      onFocus(event, ...args)
    }
    this.handleOpen()
  }

  handleChange = (event: Object, ...args: Array<any>) => {
    const { onChange } = this.props
    if (onChange) {
      onChange(event.target.value)
    }
  }

  handleOpen = () => {
    this.handleToggle(true)
  }

  handleClose = () => {
    this.handleToggle(false)
  }

  handleToggle = (open: boolean = !this.state.open) => {
    this.setState({ open, search: '' })
  }

  generateSelectHandler = (item: Object) => {
    return (event: Object) => {
      const { onChange, getItemValue } = this.props
      if (onChange) {
        onChange(getItemValue(item))
      }
      this.handleClose()
    }
  }

  getItems = () => {
    const { items } = this.props
    const { search } = this.state

    if (search.length === 0) {
      return items
    } else {
      return this.props.getSearchResults(items, search)
    }
  }
}
