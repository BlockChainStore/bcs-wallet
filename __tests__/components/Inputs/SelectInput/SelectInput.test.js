import React from 'react'
import { shallow, mount } from 'enzyme'
import SelectInput from '../../../../app/components/Inputs/SelectInput/SelectInput'

describe('SelectInput', () => {
  const props = {
    className: 'customClassName',
    value: 'some text',
    items: ['Foo', 'Bar', 'Has some text'],
    onFocus: jest.fn(),
    onChange: jest.fn()
  }

  test('should render without crashing', () => {
    const wrapper = shallow(<SelectInput {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('should render the input correctly', () => {
    const wrapper = shallow(<SelectInput {...props} />)
    expect(wrapper.dive().find('.input').props().value).toEqual(props.value)
  })

  test('should not render the dropdown by default', () => {
    const wrapper = shallow(<SelectInput {...props} />)
    expect(wrapper.dive().find('.dropdown').exists()).toEqual(false)
  })

  test('should render the dropdown when the input has focus', () => {
    const wrapper = mount(<SelectInput {...props} />)
    wrapper.find('.input').simulate('focus')
    expect(wrapper.find('.dropdown').exists()).toEqual(true)
  })

  test('should not filter the search results on initial open', () => {
    const wrapper = mount(<SelectInput {...props} value='Foo' />)
    wrapper.find('.input').simulate('focus')
    expect(wrapper.find('.dropdownItem')).toHaveLength(props.items.length)
  })
})
