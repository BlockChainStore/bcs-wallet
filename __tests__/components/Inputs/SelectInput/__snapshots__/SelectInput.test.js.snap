// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput should render without crashing 1`] = `
<clickOutside(Dropdown)
  className="selectInput customClassName"
  onClose={[Function]}
  open={false}
  renderDropdown={[Function]}
>
  <div
    className="inputContainer"
  >
    <input
      className="input"
      onChange={[Function]}
      onFocus={[Function]}
      type="text"
      value="some text"
    />
    <div
      className="afterInput"
    >
      <DropdownButton
        onToggle={[Function]}
      />
    </div>
  </div>
</clickOutside(Dropdown)>
`;
